File: ZezeJava/src/Zeze/Services/ToTypeScriptService.java
package Zeze.Services;

import Zeze.*;
import java.util.*;

public class ToTypeScriptService extends Zeze.Services.ToTypeScriptService0 {
	public CallbackOnSocketHandshakeDone CallbackWhenSocketHandshakeDone;
	public CallbackOnSocketClose CallbackWhenSocketClose;
	public CallbackOnSocketProcessInputBuffer CallbackWhenSocketProcessInputBuffer;

	public ToTypeScriptService(String name) {
		super(name);
	}


	public final void Connect(String hostNameOrAddress, int port) {
		Connect(hostNameOrAddress, port, true);
	}

//C# TO JAVA CONVERTER WARNING: There is no Java equivalent to C#'s shadowing via the 'new' keyword:
//ORIGINAL LINE: public new void Connect(string hostNameOrAddress, int port, bool autoReconnect = true)
//C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
	public final void Connect(String hostNameOrAddress, int port, boolean autoReconnect) {
		super.Connect(hostNameOrAddress, port, autoReconnect);
	}

	public final void Send(long sessionId, Puerts.ArrayBuffer buffer, int offset, int len) {
		if (super_Keyword.GetSocket(sessionId) != null) {
			super_Keyword.GetSocket(sessionId).Send(buffer.Bytes, offset, len);
		}
	}

	public final void Close(long sessionId) {
		if (super_Keyword.GetSocket(sessionId) != null) {
			super_Keyword.GetSocket(sessionId).Dispose();
		}
	}

	public final void TickUpdate() {
		HashSet<Long> handshakeTmp;
		HashSet<Long> socketCloseTmp;
		HashMap<Long, Zeze.Serialize.ByteBuffer> inputTmp;
		synchronized (this) {
			handshakeTmp = ToHandshakeDone;
			socketCloseTmp = ToSocketClose;
			inputTmp = ToBuffer;

			ToBuffer = new HashMap < Long, getZeze().Serialize.ByteBuffer>();
			ToHandshakeDone = new HashSet<Long>();
			ToSocketClose = new HashSet<Long>();
		}

		for (var e : socketCloseTmp) {
			this.CallbackWhenSocketClose.invoke(e);
		}

		for (var e : handshakeTmp) {
			this.CallbackWhenSocketHandshakeDone.invoke(e);
		}

		for (var e : inputTmp.entrySet()) {
			this.CallbackWhenSocketProcessInputBuffer.invoke(e.getKey(), new Puerts.ArrayBuffer(e.getValue().Bytes), e.getValue().ReadIndex, e.getValue().Size);
		}
	}
}
//#endif

File: confcs/Zeze/Services/ToTypeScriptService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Zeze.Net;
using Zeze.Serialize;

namespace Zeze.Services
{
    public class ToTypeScriptService0 : HandshakeClient
    {
        public ToTypeScriptService0(string name) : base(name, (Application)null)
        {
        }

        public override void OnSocketProcessInputBuffer(AsyncSocket so, ByteBuffer input)
        {
            if (so.IsHandshakeDone)
            {
                AppendInputBuffer(so.SessionId, input);
                input.ReadIndex = input.WriteIndex;
            }
            else
            {
                base.OnSocketProcessInputBuffer(so, input);
            }
        }

        public override void OnSocketClose(AsyncSocket so, Exception e)
        {
            SetSocketClose(so.SessionId);
            base.OnSocketClose(so, e);
        }

        public override void OnHandshakeDone(AsyncSocket sender)
        {
            sender.IsHandshakeDone = true;
            SetHandshakeDone(sender.SessionId);
        }

        protected Dictionary<long, ByteBuffer> ToBuffer = new Dictionary<long, ByteBuffer>();
        protected HashSet<long> ToHandshakeDone = new HashSet<long>();
        protected HashSet<long> ToSocketClose = new HashSet<long>();

        internal void SetHandshakeDone(long socketSessionId)
        {
            lock (this)
            {
                ToHandshakeDone.Add(socketSessionId);
            }
        }

        internal void SetSocketClose(long socketSessionId)
        {
            lock (this)
            {
                ToSocketClose.Add(socketSessionId);
            }
        }

        internal void AppendInputBuffer(long socketSessionId, ByteBuffer buffer)
        {
            lock (this)
            {
                if (ToBuffer.TryGetValue(socketSessionId, out var exist))
                {
                    exist.Append(buffer.Bytes, buffer.ReadIndex, buffer.Size);
                    return;
                }
                ByteBuffer newBuffer = ByteBuffer.Allocate();
                ToBuffer.Add(socketSessionId, newBuffer);
                newBuffer.Append(buffer.Bytes, buffer.ReadIndex, buffer.Size);
            }
        }
    }
}

#if USE_UNITY_PUERTS
// 下面这个类是真正开放给ts用的，使用Puerts绑定，需要Puerts支持。
// 使用的时候拷贝下面的代码到你自己的ToTypeScriptService.cs文件。
// 并且在Puerts.Binding里面增加 typeof 绑定到ts。

public delegate void CallbackOnSocketHandshakeDone(long sessionId);
public delegate void CallbackOnSocketClose(long sessionId);
public delegate void CallbackOnSocketProcessInputBuffer(long sessionId, Puerts.ArrayBuffer buffer, int offset, int len);

public class ToTypeScriptService : Zeze.Services.ToTypeScriptService0
{
    public CallbackOnSocketHandshakeDone CallbackWhenSocketHandshakeDone;
    public CallbackOnSocketClose CallbackWhenSocketClose;
    public CallbackOnSocketProcessInputBuffer CallbackWhenSocketProcessInputBuffer;

    public ToTypeScriptService(string name) : base(name)
    {
    }

    public new void Connect(string hostNameOrAddress, int port, bool autoReconnect = true)
    {
        base.Connect(hostNameOrAddress, port, autoReconnect);
    }

    public void Send(long sessionId, Puerts.ArrayBuffer buffer, int offset, int len)
    {
        base.GetSocket(sessionId)?.Send(buffer.Bytes, offset, len);
    }
        
    public void Close(long sessionId)
    {
        base.GetSocket(sessionId)?.Dispose();
    }

    public void TickUpdate()
    {
        System.Collections.Generic.HashSet<long> handshakeTmp;
        System.Collections.Generic.HashSet<long> socketCloseTmp;
        System.Collections.Generic.Dictionary<long, Zeze.Serialize.ByteBuffer> inputTmp;
        lock (this)
        {
            handshakeTmp = ToHandshakeDone;
            socketCloseTmp = ToSocketClose;
            inputTmp = ToBuffer;

            ToBuffer = new System.Collections.Generic.Dictionary<long, Zeze.Serialize.ByteBuffer>();
            ToHandshakeDone = new System.Collections.Generic.HashSet<long>();
            ToSocketClose = new System.Collections.Generic.HashSet<long>();
        }

        foreach (var e in socketCloseTmp)
        {
            this.CallbackWhenSocketClose(e);
        }

        foreach (var e in handshakeTmp)
        {
            this.CallbackWhenSocketHandshakeDone(e);
        }

        foreach (var e in inputTmp)
        {
            this.CallbackWhenSocketProcessInputBuffer(e.Key, new Puerts.ArrayBuffer(e.Value.Bytes), e.Value.ReadIndex, e.Value.Size);
        }
    }
}

#endif